core/dictgen/src/rootcling_impl.cxx

```diff
@@ -4233,9 +4253,46 @@ int RootClingMain(int argc,
 
    ROOT::TMetaUtils::SetPathsForRelocatability(clingArgs);
 
+   // FIXME: This line is from TModuleGenerator, but we can't reuse this code
+   // at this point because TModuleGenerator needs a CompilerInstance (and we
+   // currently create the arguments for creating said CompilerInstance).
+   bool isPCH = (dictpathname == "allDict.cxx");
+   StringRef moduleName;
+   // Adding -fmodules to the args will break lexing with __CINT__ defined,
+   // and we actually do lex with __CINT__ and reuse this variable later,
+   // we have to copy it now.
+   auto clingArgsInterpreter = clingArgs;
+   if (!isPCH && getenv("ROOT_MODULES")) {
+      // We just pass -fmodules, the CIFactory will do the rest and configure
+      // clang correctly once it sees this flag.
+      clingArgsInterpreter.push_back("-fmodules");
+
+      // Specify the module name that we can lookup the module in the modulemap.
+      std::string outputFile = llvm::sys::path::stem(sharedLibraryPathName).str();
+      // Try to get the module name in the modulemap based on the filepath.
+      moduleName = llvm::sys::path::filename(outputFile);
+      moduleName.consume_front("lib");
+      moduleName.consume_back("_rdict.pcm");
+
+      clingArgsInterpreter.push_back("-fmodule-name");
+      clingArgsInterpreter.push_back(moduleName.str());
+
+      // Set the C++ modules output directory to the directory where we generate
+      // the shared library.
+      clingArgsInterpreter.push_back("-fmodules-cache-path=" +
+                                     llvm::sys::path::parent_path(sharedLibraryPathName).str());
+
+      // If the user has passed ROOT_MODULES with the value 'DEBUG', then we
+      // enable remarks in clang for building on-demand modules. This is useful
+      // to figure out when and why on-demand modules are built by clang.
+      if (llvm::StringRef(getenv("ROOT_MODULES")) == "DEBUG") {
```

Comment:
Isn't this check redundant? We will check it in the CIFactory again and add the flag iiuc.

Reply:
Yes, it's an artifact when I tried to move out ROOT-specific code from CIFactory to rootcling. FIxed!